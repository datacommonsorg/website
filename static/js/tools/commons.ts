/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Interface class to filter stats vars. This should be
 * implemented as a concrete class based on specific use case.
 */
interface StatsVarFilterInterface {
  isValid(statsVar: string): boolean;
}

/**
 * StatsVar filter class for timeline tool.
 */
class TimelineStatsVarFilter implements StatsVarFilterInterface {
  statsVars: Set<string>;
  isValid(statsVar: string): boolean {
    return this.statsVars.has(statsVar);
  }
  constructor(statsVars: Set<string>) {
    this.statsVars = statsVars;
  }
}

/**
 * StatsVar filter for choropleth tool.
 */
class ChoroplethStatsVarFilter implements StatsVarFilterInterface {
  isValid(statsVar: string): boolean {
    // TODO(shifucun): add the actual logic.
    return !!statsVar;
  }
}

/**
 * StatsVar filter that does not filter and allows all non-null stats vars.
 */
class NoopStatsVarFilter implements StatsVarFilterInterface {
  isValid(statsVar: string): boolean {
    return !!statsVar;
  }
}

export {
  ChoroplethStatsVarFilter,
  NoopStatsVarFilter,
  StatsVarFilterInterface,
  TimelineStatsVarFilter,
};
