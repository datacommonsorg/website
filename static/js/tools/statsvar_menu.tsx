import React, { Component } from "react";
import axios from "axios";
import hierarchy from "../../data/hierarchy_top.json";
import { updateUrl } from "./timeline_util";

const jsonPath = "../../data/hierarchy_statsvar.json";

interface NodePropType {
  l: string; // label
  c: number; // count
  cd: NodePropType[]; // children
  t: string; // type
  sv: string;
  statsVarPaths: number[][];
  nodePath: number[];
  statsVarValid: Set<string>;
  filter: boolean;
  idx: number;
  setName: (statsVarId: string, statsVarName: string) => void;
}

interface NodeStateType {
  checked: boolean;
  expanded: boolean;
  statsVarPaths: number[][];
}

class Node extends Component<NodePropType, NodeStateType> {
  constructor(props: NodePropType) {
    super(props);
    this._handleCheckboxClick = this._handleCheckboxClick.bind(this);
    this._handleExpandClick = this._handleExpandClick.bind(this);
    this._handleHashChange = this._handleHashChange.bind(this);
    this.canExpand = this.canExpand.bind(this);
    this.isValidStatsVar = this.isValidStatsVar.bind(this);
    this.state = {
      checked: false,
      expanded: false,
      statsVarPaths: [[]],
    };
  }

  componentDidUpdate(prevProps) {
    if (this.props.statsVarPaths !== prevProps.statsVarPaths) {
      this._handleHashChange();
    }
  }
  componentDidMount() {
    this._handleHashChange();
  }

  public render = (): JSX.Element => {
    let child: JSX.Element[];
    const isValidStatsVar = this.isValidStatsVar();
    const canExpand = this.canExpand();
    if (canExpand && this.state.expanded) {
      child = this.props.cd.map((item, index) => {
        return (
          <Node
            l={item.l}
            cd={item.cd}
            c={item.c}
            t={item.t}
            sv={item.sv}
            statsVarPaths={this.state.statsVarPaths}
            key={this.props.l + index}
            statsVarValid={this.props.statsVarValid}
            filter={this.props.filter}
            idx={index}
            setName={this.props.setName}
            nodePath={[...this.props.nodePath, index]}
          ></Node>
        );
      });
    }

    return (
      // render the node only if it is a valid value node
      // or it is a property node that can be expanded
      (isValidStatsVar || (this.props.t !== "v" && canExpand)) && (
        <ul className="noborder">
          <li className="value" id={this.props.l}>
            <span>
              <a className="value-link">
                {this.props.l + "  "}
                <sup>{this.props.c !== 0 && "(" + this.props.c + ")"}</sup>
                {isValidStatsVar && (
                  <button
                    className={
                      this.state.checked ? "checkbox checked" : "checkbox"
                    }
                    onClick={this._handleCheckboxClick}
                  />
                )}
                {canExpand && (
                  <img
                    className={
                      this.state.expanded
                        ? "right-caret transform-up"
                        : "right-caret"
                    }
                    src="/images/right-caret-light.png"
                    onClick={this._handleExpandClick}
                  />
                )}
              </a>
            </span>
            {child}
          </li>
        </ul>
      )
    );
  };
  private _handleCheckboxClick = (): void => {
    this.setState({
      checked: !this.state.checked,
    });
    updateUrl({
      statsVar: {
        statsVar: this.props.sv + "," + this.props.nodePath.join(","),
        shouldAdd: !this.state.checked,
      },
    });
  };

  private _handleExpandClick = (): void => {
    this.setState({
      expanded: !this.state.expanded,
    });
  };

  private _handleHashChange() {
    const statsVarPathNext = [];
    let check = false;
    let expand = false;
    for (const statsVarPath of this.props.statsVarPaths) {
      if (statsVarPath && statsVarPath[0] === this.props.idx) {
        if (statsVarPath.length === 1) {
          check = true;
          this.props.setName(this.props.sv, this.props.l);
        } else {
          expand = true;
          statsVarPathNext.push(statsVarPath.slice(1));
        }
      }
    }
    this.setState({
      checked: check,
      expanded: expand,
      statsVarPaths: statsVarPathNext,
    });
  }

  private isValidStatsVar() {
    // the node is valid statsvar node if it is a value node,
    // and the statsvar is available or not filtered.
    return (
      this.props.t === "v" &&
      (!this.props.filter || this.props.statsVarValid.has(this.props.sv))
    );
  }

  private canExpand() {
    if (this.props.t === "p") {
      // a property node can be expanded if it has >= 1 children
      return this.hasChild(this.props.cd);
    } else if (this.props.t === "c") {
      // the top level node is expandable if has valid value node
      // or valid property node
      let valid = false;
      this.props.cd.map((item) => {
        if (
          item.t === "v" &&
          (!this.props.filter || this.props.statsVarValid.has(item.sv))
        ) {
          valid = true; // valid value node
        } else if (this.hasChild(item.cd)) {
          valid = true; // valid property node
        }
      });
      return valid;
    } else {
      // a value node is expandable if it has valid property node
      let valid = false;
      if (this.props.cd) {
        this.props.cd.map((item) => {
          if (this.hasChild(item.cd)) {
            valid = true;
          }
        });
      }
      return valid;
    }
  }

  private hasChild(children) {
    // return true if a property node has valid children
    let valid = false;
    if (children && children.length !== 0) {
      children.map((item) => {
        if (!this.props.filter || this.props.statsVarValid.has(item.sv)) {
          valid = true;
        }
      });
    }
    return valid;
  }
}

interface MenuPropType {
  search: boolean;
  statsVarPaths: number[][];
  statsVarValid: Set<string>;
  filter: boolean;
  setName: (statsVarId: string, statsVarName: string) => void;
}
interface MenuStateType {
  menuJson: [{}];
}
class Menu extends Component<MenuPropType, MenuStateType> {
  constructor(props) {
    super(props);
    this.state = {
      menuJson: [hierarchy],
    };
  }
  render() {
    return (
      <div id="drill">
        <div className="noedge">
          {this.state.menuJson.map((vertical, index1) => {
            return Object.keys(vertical).map((key, index) => {
              const item = vertical[key];
              return (
                item.cd.length !== 0 && (
                  <Node
                    l={item.l}
                    cd={item.cd}
                    c={item.c}
                    t="c"
                    sv={item.sv}
                    key={index1 + "," + index}
                    statsVarPaths={this.props.statsVarPaths}
                    statsVarValid={this.props.statsVarValid}
                    filter={this.props.filter}
                    idx={index}
                    setName={this.props.setName}
                    nodePath={[index]}
                  ></Node>
                )
              );
            });
          })}
        </div>
      </div>
    );
  }
  componentDidMount() {
    axios.get(jsonPath).then((resp) => {
      this.setState({
        menuJson: [resp.data],
      });
    });
  }
}

export { Menu };
