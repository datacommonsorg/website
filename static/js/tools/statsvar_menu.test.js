/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { Menu } from "./statsvar_menu";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import pretty from "pretty";
import { TimelineStatsVarFilter } from "./commons";

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("filtering the menu", () => {
  // input: one valid statsVar selected
  // expecting: menu filtered, node Demographics expanded, node Count_Person checked
  act(() => {
    render(
      <Menu
        selectedNodePaths={[["0", "0"]]}
        setStatsVarTitle={jest.fn()}
        addStatsVar={jest.fn()}
        removeStatsVar={jest.fn()}
        statsVarFilter={new TimelineStatsVarFilter(new Set(["Count_Person"]))}
      />,
      container
    );
  });
  expect(pretty(container.innerHTML)).toMatchSnapshot();

  // input: non-valid statsVar, expecting: empty menu
  act(() => {
    render(
      <Menu
        selectedNodePaths={[["0", "0"]]}
        setStatsVarTitle={jest.fn()}
        addStatsVar={jest.fn()}
        removeStatsVar={jest.fn()}
        statsVarFilter={new TimelineStatsVarFilter(new Set([]))}
      />,
      container
    );
  });
  expect(pretty(container.innerHTML)).toMatchSnapshot();

  // input: selected statsVar Count_Person is not valid
  // expecting: Medium Age expanded, Count_Person not shown
  act(() => {
    render(
      <Menu
        selectedNodePaths={[["0", "0"]]}
        setStatsVarTitle={jest.fn()}
        addStatsVar={jest.fn()}
        removeStatsVar={jest.fn()}
        statsVarFilter={
          new TimelineStatsVarFilter(new Set(["Median_Age_Person"]))
        }
      />,
      container
    );
  });
  expect(pretty(container.innerHTML)).toMatchSnapshot();
});

// Todo(Lijuan): add test for setting statsVar titles
