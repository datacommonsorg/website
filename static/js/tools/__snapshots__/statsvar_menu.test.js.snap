// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filtering the menu 1`] = `
"<div id=\\"drill\\">
  <div class=\\"noedge\\">
    <ul class=\\"noborder\\">
      <li class=\\"value\\" id=\\"Demographics\\"><span><a class=\\"value-link\\">Demographics  <sup>(1787)</sup><img class=\\"right-caret transform-up\\" src=\\"/images/right-caret-light.png\\"></a></span>
        <ul class=\\"noborder\\">
          <li class=\\"value\\" id=\\"Population\\"><span><a class=\\"value-link\\">Population  <sup>(1)</sup><button class=\\"checkbox checked\\"></button></a></span></li>
        </ul>
      </li>
    </ul>
  </div>
</div>"
`;

exports[`filtering the menu 2`] = `
"<div id=\\"drill\\">
  <div class=\\"noedge\\"></div>
</div>"
`;

exports[`filtering the menu 3`] = `
"<div id=\\"drill\\">
  <div class=\\"noedge\\">
    <ul class=\\"noborder\\">
      <li class=\\"value\\" id=\\"Demographics\\"><span><a class=\\"value-link\\">Demographics  <sup>(1787)</sup><img class=\\"right-caret transform-up\\" src=\\"/images/right-caret-light.png\\"></a></span>
        <ul class=\\"noborder\\">
          <li class=\\"value\\" id=\\"Median age\\"><span><a class=\\"value-link\\">Median age  <sup>(1)</sup><button class=\\"checkbox\\"></button></a></span></li>
        </ul>
      </li>
    </ul>
  </div>
</div>"
`;
